### ISSUE TITLE
Problems with Mojave?

### ISSUE THREAD

**Issue by progrium** [JXA-Cookbook/JXA-Cookbook#44]
Since I upgraded to macOS 10.14, `$.NSMakeRect` now always throws `NSInvalidArgumentException` with reason: `NSGetSizeAndAlignment(): unsupported type encoding spec 'G' at 'GPoint}"size"{CGSize}}' in '{CGPoint}"size"{CGSize}}'` ... wat?!

Anybody else run into this? Ideas on what's going on?

Program:
```
#!/usr/bin/env osascript -l JavaScript
ObjC.import("Cocoa");

function run(argv) {
    var r = $.NSMakeRect(0, 0, 100, 100);
}
```
Result:
```
2018-10-16 11:16:52.101 osascript[53689:2461150] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'NSGetSizeAndAlignment(): unsupported type encoding spec 'G' at 'GPoint}"size"{CGSize}}' in '{CGPoint}"size"{CGSize}}''
*** First throw call stack:
(
        0   CoreFoundation                      0x00007fff2f762571 __exceptionPreprocess + 256
        1   libobjc.A.dylib                     0x00007fff5b70fefb objc_exception_throw + 48
        2   CoreFoundation                      0x00007fff2f6b4bf0 __NSGetSizeAndAlignment + 1093
        3   CoreFoundation                      0x00007fff2f6b4a60 __NSGetSizeAndAlignment + 693
        4   CoreFoundation                      0x00007fff2f6c17d5 NSGetSizeAndAlignment + 33
        5   JavaScriptAppleEvents               0x00007fff4db298ba JSOCObjCTypeIterateTypes + 136
        6   JavaScriptAppleEvents               0x00007fff4db299ae JSOCObjCTypeIterateStructElements + 131
        7   JavaScriptAppleEvents               0x00007fff4db490a7 +[JSOCType typeWithObjCType:] + 457
        8   JavaScriptAppleEvents               0x00007fff4db276af -[JSOCBridgeSupport typeForArgumentInfo:fallbackObjCType:] + 475
        9   JavaScriptAppleEvents               0x00007fff4db27751 -[JSOCBridgeSupport typeForFunctionInfo:isBlock:] + 99
        10  JavaScriptAppleEvents               0x00007fff4db282ac -[JSOCBridgeSupport functionPointerForFunctionWithName:] + 89
        11  JavaScriptAppleEvents               0x00007fff4db22f92 __JSOCFunctionConstructorBlock_block_invoke + 304
        12  CoreFoundation                      0x00007fff2f6b80ec __invoking___ + 140
        13  CoreFoundation                      0x00007fff2f6b7fb7 -[NSInvocation invoke] + 311
        14  JavaScriptCore                      0x00007fff32d0cde8 _ZN3JSC24ObjCCallbackFunctionImpl4callEP9JSContextP13OpaqueJSValuemPKPKS3_PS6_ + 440
        15  JavaScriptCore                      0x00007fff32d0c81c _ZN3JSCL34objCCallbackFunctionCallAsFunctionEPK15OpaqueJSContextP13OpaqueJSValueS4_mPKPKS3_PS6_ + 236
        16  JavaScriptCore                      0x00007fff32d0c0ec _ZN3JSC19APICallbackFunction4callINS_20ObjCCallbackFunctionEEExPNS_9ExecStateE + 492
        17  JavaScriptCore                      0x00007fff32cf4d63 vmEntryToNative + 241
        18  JavaScriptCore                      0x00007fff32b1e7fe _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 542
        19  JavaScriptCore                      0x00007fff33546051 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 177
        20  JavaScriptCore                      0x00007fff32b1e474 JSObjectCallAsFunction + 468
        21  JavaScriptCore                      0x00007fff32c84763 -[JSValue callWithArguments:] + 291
        22  JavaScriptAppleEvents               0x00007fff4db30215 __JSOCDollarObject_block_invoke_2 + 357
        23  CoreFoundation                      0x00007fff2f6b80ec __invoking___ + 140
        24  CoreFoundation                      0x00007fff2f6b7fb7 -[NSInvocation invoke] + 311
        25  JavaScriptCore                      0x00007fff32d0cde8 _ZN3JSC24ObjCCallbackFunctionImpl4callEP9JSContextP13OpaqueJSValuemPKPKS3_PS6_ + 440
        26  JavaScriptCore                      0x00007fff32d0c81c _ZN3JSCL34objCCallbackFunctionCallAsFunctionEPK15OpaqueJSContextP13OpaqueJSValueS4_mPKPKS3_PS6_ + 236
        27  JavaScriptCore                      0x00007fff32d0c0ec _ZN3JSC19APICallbackFunction4callINS_20ObjCCallbackFunctionEEExPNS_9ExecStateE + 492
        28  JavaScriptCore                      0x00007fff32cf4d63 vmEntryToNative + 241
        29  JavaScriptCore                      0x00007fff32b1e7fe _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 542
        30  JavaScriptCore                      0x00007fff33546051 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 177
        31  JavaScriptCore                      0x00007fff32b1e474 JSObjectCallAsFunction + 468
        32  JavaScriptAppleEvents               0x00007fff4db2fd0e GetProperty + 309
        33  JavaScriptCore                      0x00007fff32d27361 _ZN3JSC16JSCallbackObjectINS_20JSDestructibleObjectEE14callbackGetterEPNS_9ExecStateExNS_12PropertyNameE + 225
        34  JavaScriptCore                      0x00007fff32ca1876 llint_slow_path_get_by_id + 5302
        35  JavaScriptCore                      0x00007fff32cf7a73 llint_entry + 11449
        36  JavaScriptCore                      0x00007fff32cf4c45 vmEntryToJavaScript + 235
        37  JavaScriptCore                      0x00007fff32b1e7b6 _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 470
        38  JavaScriptCore                      0x00007fff33546051 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 177
        39  JavaScriptCore                      0x00007fff32b1e474 JSObjectCallAsFunction + 468
        40  JavaScriptCore                      0x00007fff32c84763 -[JSValue callWithArguments:] + 291
        41  JavaScriptOSA                       0x00007fff4db71927 -[JSStorage(JSProcedures) handleEvent:inContext:modeFlags:resultingResultID:] + 3551
        42  CoreFoundation                      0x00007fff2f6b80ec __invoking___ + 140
        43  CoreFoundation                      0x00007fff2f6b7fb7 -[NSInvocation invoke] + 311
        44  JavaScript                          0x0000000117c517bc JavaScriptComponent + 1140
        45  OpenScripting                       0x00007fff2ece12da OSAExecuteEvent + 52
        46  osascript                           0x000000010c73b565 osascript + 9573
        47  libdyld.dylib                       0x00007fff5c7d6085 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
Abort trap: 6
```

**Comment by gyratorycircus** [comment]
I've encountered this as well, and I haven't been able to work around it in any way. It seems that anything related to `NSRect` throws the same exception. Can't even access the frame of a view without encountering it:

    let view = $.NSView.alloc.init;
    let rect = view.frame;

**Comment by Tatsh** [comment]
This seems like a bug that should be reported to Apple.

**Comment by SirVer** [comment]
Did somebody find a workaround? I have the same problem since after upgrading to Mojave.

**Comment by Tatsh** [comment]
I reported this to Apple so we'll see. I figure your workaround of setting `bounds` can be used for now?

**Comment by SirVer** [comment]
I unblocked myself by reimplementing the script I was using for moving windows around in Rust, directly calling into the objective C libraries. That works for me for the moment. Thanks for reporting the bug to Apple, I think that is the exactly right call.

**Comment by rcarmo** [comment]
I'm guessing this will take a while to be fixed, but any workarounds that I can use in JXA itself? What's the mention to "setting  `bounds`" above? Is it setting the window size post-creation?

**Comment by piersg** [comment]
This has been annoying me for some time and is blocking a Mojave upgrade on quite a few systems where we have an extremely useful JS/GUI script.

I too have reported it to Apple and everyone with a developer account should too - the more reports the more chance they'll actually give it some sort of priority.

**Workaround**

Looking at the crash report it just looks like - possibly - someone broke the JavaScriptAppleEvents bridge with a typo or bad compiler flag or something. It also seems pretty unlikely any actual features got added (or that work of any sort got done on it), so for fun I copied the framework from a High Sierra machine over the top of the Mojave one - they're both v. 1.0 ©2013 Apple but they are different sizes/file counts.

And my script runs again! The framework is in /System/Library/PrivateFrameworks/

Caveat: barely tested, use at own risk, doing it en masse might make Apple mad at you.

**Comment by Tatsh** [comment]
> And my script runs again! The framework is in /System/Library/PrivateFrameworks/

I presume you disabled SIP?

**Comment by piersg** [comment]
I was wondering if I'd have to but it let me with just an admin password. It's on a VM which I'm 98% sure I haven't de-SIPPed.

So, no I don't think so

**Comment by paul-go** [comment]
Can confirm that on a fresh install of Mojave, (SIP not disabled), attempting to alter /System/Library/PrivateFrameworks/JavaScriptAppleEvents.framework results in an `operation not permitted`, even with sudo.

Maybe somebody can check to see if the issue still persists on Catalina?

**Comment by Tatsh** [comment]
May be out of date with Catalina. Closing in 30 days.

**Comment by SirVer** [comment]
Still a problem with 10.15.2, running the script from OP results in a very similar stack trace and the same error message:

~~~
2020-03-09 09:49:25.871 osascript[96865:25463852] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'NSGetSizeAndAlignment(): unsupported type encoding spec 'G' at 'GPoint}"size"{CGSize}}' in '{CGPoint}"size"{CGSize}}''
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff2c6a738b __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff629d5552 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff2c60a1fc __NSGetSizeAndAlignment + 1125
	3   CoreFoundation                      0x00007fff2c60a068 __NSGetSizeAndAlignment + 721
	4   CoreFoundation                      0x00007fff2c615db1 NSGetSizeAndAlignment + 33
	5   JavaScriptAppleEvents               0x00007fff50f82188 JSOCObjCTypeIterateTypes + 136
	6   JavaScriptAppleEvents               0x00007fff50f82272 JSOCObjCTypeIterateStructElements + 121
	7   JavaScriptAppleEvents               0x00007fff50fa1363 +[JSOCType typeWithObjCType:] + 447
	8   JavaScriptAppleEvents               0x00007fff50f8001a -[JSOCBridgeSupport typeForArgumentInfo:fallbackObjCType:] + 504
	9   JavaScriptAppleEvents               0x00007fff50f80086 -[JSOCBridgeSupport typeForFunctionInfo:isBlock:] + 103
	10  JavaScriptAppleEvents               0x00007fff50f80b9d -[JSOCBridgeSupport functionPointerForFunctionWithName:] + 89
	11  JavaScriptAppleEvents               0x00007fff50f7b9d4 __JSOCFunctionConstructorBlock_block_invoke + 304
	12  CoreFoundation                      0x00007fff2c60cefc __invoking___ + 140
	13  CoreFoundation                      0x00007fff2c60cd98 -[NSInvocation invoke] + 305
	14  JavaScriptCore                      0x00007fff30222a5b _ZN3JSC24ObjCCallbackFunctionImpl4callEP9JSContextP13OpaqueJSValuemPKPKS3_PS6_ + 443
	15  JavaScriptCore                      0x00007fff30222484 _ZN3JSCL34objCCallbackFunctionCallAsFunctionEPK15OpaqueJSContextP13OpaqueJSValueS4_mPKPKS3_PS6_ + 244
	16  JavaScriptCore                      0x00007fff30221daf _ZN3JSC19APICallbackFunction4callINS_20ObjCCallbackFunctionEEExPNS_9ExecStateE + 511
	17  JavaScriptCore                      0x00007fff301e40d6 vmEntryToNative + 206
	18  JavaScriptCore                      0x00007fff2fed01e7 _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 487
	19  JavaScriptCore                      0x00007fff30981e74 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 196
	20  JavaScriptCore                      0x00007fff2fecfe98 JSObjectCallAsFunction + 488
	21  JavaScriptCore                      0x00007fff3002179c -[JSValue callWithArguments:] + 252
	22  JavaScriptAppleEvents               0x00007fff50f889b2 __JSOCDollarObject_block_invoke_2 + 357
	23  CoreFoundation                      0x00007fff2c60cefc __invoking___ + 140
	24  CoreFoundation                      0x00007fff2c60cd98 -[NSInvocation invoke] + 305
	25  JavaScriptCore                      0x00007fff30222a5b _ZN3JSC24ObjCCallbackFunctionImpl4callEP9JSContextP13OpaqueJSValuemPKPKS3_PS6_ + 443
	26  JavaScriptCore                      0x00007fff30222484 _ZN3JSCL34objCCallbackFunctionCallAsFunctionEPK15OpaqueJSContextP13OpaqueJSValueS4_mPKPKS3_PS6_ + 244
	27  JavaScriptCore                      0x00007fff30221daf _ZN3JSC19APICallbackFunction4callINS_20ObjCCallbackFunctionEEExPNS_9ExecStateE + 511
	28  JavaScriptCore                      0x00007fff301e40d6 vmEntryToNative + 206
	29  JavaScriptCore                      0x00007fff2fed01e7 _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 487
	30  JavaScriptCore                      0x00007fff30981e74 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 196
	31  JavaScriptCore                      0x00007fff2fecfe98 JSObjectCallAsFunction + 488
	32  JavaScriptAppleEvents               0x00007fff50f884ac GetProperty + 309
	33  JavaScriptCore                      0x00007fff3022d481 _ZN3JSC16JSCallbackObjectINS_20JSDestructibleObjectEE14callbackGetterEPNS_9ExecStateExNS_12PropertyNameE + 225
	34  JavaScriptCore                      0x00007fff3003e94a llint_slow_path_get_by_id + 3274
	35  JavaScriptCore                      0x00007fff301ed6ce llint_entry + 38261
	36  JavaScriptCore                      0x00007fff301e3faf vmEntryToJavaScript + 200
	37  JavaScriptCore                      0x00007fff2fed01a0 _ZN3JSC11Interpreter11executeCallEPNS_9ExecStateEPNS_8JSObjectENS_8CallTypeERKNS_8CallDataENS_7JSValueERKNS_7ArgListE + 416
	38  JavaScriptCore                      0x00007fff30981e74 _ZN3JSC12profiledCallEPNS_9ExecStateENS_15ProfilingReasonENS_7JSValueENS_8CallTypeERKNS_8CallDataES3_RKNS_7ArgListE + 196
	39  JavaScriptCore                      0x00007fff2fecfe98 JSObjectCallAsFunction + 488
	40  JavaScriptCore                      0x00007fff3002179c -[JSValue callWithArguments:] + 252
	41  JavaScriptOSA                       0x00007fff50fc9ad8 -[JSStorage(JSProcedures) handleEvent:inContext:modeFlags:resultingResultID:] + 3483
	42  CoreFoundation                      0x00007fff2c60cefc __invoking___ + 140
	43  CoreFoundation                      0x00007fff2c60cd98 -[NSInvocation invoke] + 305
	44  JavaScript                          0x0000000108b5d7c0 JavaScriptComponent + 1140
	45  OpenScripting                       0x00007fff2b4533f2 OSAExecuteEvent + 50
	46  osascript                           0x00000001089c55a3 osascript + 9635
	47  libdyld.dylib                       0x00007fff63d377fd start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
~~~

**Comment by 4r7d3c0** [comment]
omg apple how could you break _this_ goodness. is there absolutely no way to create a window...

**Comment by Tatsh** [comment]
I meant to close this 4 months ago.

**Comment by 4r7d3c0** [comment]
why close the issue is very much an issue. you should keep this open. it was hard already to find it, let others see that's it's not resolved if they stumble upon it. i want to post a workaround tomorrow after i fix this (having spent a day bashing my head against keyboard, I've got a clue but i need to download xcode tonight). please keep this open until the wiki is properly updated to make people know that they can't open windows because of apple, and how they can (if my solution works). thanks

**Comment by wh1pch81n** [comment]
In catalina merely calling `var rect = $.NSMakeRect(0,0,50,50)`
will cause the crash.

Also...
Upgraded to Big sur beta.  You can't even `$.NSView.alloc.init` now.

**Comment by chrillek** [comment]
The script does not throw an error or exception on Mac OS 12.0.1 (Monterey). So maybe (!) Apple did fix something?

**Comment by piersg** [comment]
+1 on my script not crashing in Monterey. (Don't think I want to upgrade the fleet to a .0.1 macOS, but maybe soon ...)

**Comment by chrillek** [comment]
Hi,

you mentioned an "extremely usefule JS/GUI script" in one of your posts. Do you have any pointers on how to implement GUIs in JXA? I found some older stuff, but it is badly documented (if at all) and I'm not sure if/how it is even possible to build anything more complicated than Apple's displaySomething methods.

tx in advance
Christian

**Comment by piersg** [comment]
Hi .. is there any way to take these offline (that isn't put your email in a github comment)?

Anyway, the script is something I made on my employer's time so it's a little proprietary. Most everything I learned came from Tyler Gaw's https://github.com/tylergaw/js-osx-app-examples. I have implemented an NSTable, which I did with a mix of the aforementioned examples and Apple sample Objective-C code for NSTable.

I just might one day roll that into an example and post it here, but I couldn't promise it
